src/middleware/authMiddleware.ts

```diff
@@ -0,0 +1,205 @@
+}
+
+/**
+ * Generates a JWT to be used as a token
+ * 
+ * @param uid the uuid of the user
+ * @param type the type of the token - access, refresh, or reset
+ * 
+ * @returns an Encode Result object containing the token, issue date and expiry date
+ */
+export const encodeJWT = (uid: string, type: string): EncodeResult => {
+    const algorithm: TAlgorithm = 'HS512';
+    const issueDate = Date.now();
+    let expiryDate;
+
+    if(type === 'access') {
+        expiryDate = Date.now() + ACCESS_EXPIRY;
+    } else if(type === 'reset') {
+        expiryDate = Date.now() + RESET_EXPIRY;
+    } else {
+        expiryDate = undefined; // Never expire
+    }
+    
+    const session: Session = {
+        id: uid,
+        issueDate: issueDate,
+        expiryDate: expiryDate
+    };
+    return {
+        token: encode(session, SECRET_KEY, algorithm),
+        issueDate: issueDate,
+        expiryDate: expiryDate
+    };
+};
+
+/**
+ * Decodes a given token and checks if it is valid
+ * 
+ * @param token the JWT to decode
+ * 
+ * @returns DecodeResult object containing: Whether the JWT is valid and the session (if it was valid) or an error message (if invalid)
+ */
+export const decodeJWT = (token: string): DecodeResult => {
+    const algorithm: TAlgorithm = 'HS512';
+    try {
+        const decoded = decode(token, SECRET_KEY, false, algorithm);
+        return {
+            valid: true,
+            session: decoded
+        };
+    } catch (e) {
+        return {
+            valid: false,
+            error: e.message
+        };
+    }
+};
+
+/**
+ * Handles the creation and management of new tokens. Generates a JWT, and saves it to the appropriate table
+ * 
+ * @param uid the uid of the user
+ * @param type the type of the token - access, refresh, or reset
+ * 
+ * @returns the EncodeResult returned by encodeJWT()
+ */
+export const generateToken = async (uid: string, type: string): Promise<EncodeResult> => {
+    const result: EncodeResult = encodeJWT(uid, type);
+    if(type === 'access') {
+        const accessTokenRepo = getManager().getRepository(AccessToken);
+        await accessTokenRepo.save(accessTokenRepo.create({token: result.token, uuid: uid}));
+    } else if(type === 'refresh') {
+        const refreshTokenRepo = getManager().getRepository(RefreshToken);
+        await refreshTokenRepo.save(refreshTokenRepo.create({token: result.token, uuid: uid}));
+    } else if(type === 'reset') {
+        const resetRepo = getManager().getRepository(PasswordReset);
+        await resetRepo.save(resetRepo.create({token: result.token, uuid: uid}));
+    }
+    
+    return result;
+}
+
+/**
+ * Verifies if a token a valid or not
+ * 
+ * @param token the token
+ * @param type the expected type of the token
+ * 
+ * @returns VerifyResult object, with result of the decoding and the uid of the token owner if the token was valid
+ */
+export const verifyToken = async (token: string, type: string): Promise<VerifyResult> => {
```

Comment:
Would it make sense to extract a Type to represent the `type` parameter?
e.g. something like 
```ts
type TokenType = "access" | "refresh" | "reset";
```